syne_tune/callbacks/hyperband_remove_checkpoints_callback.py

```diff
@@ -0,0 +1,522 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+from typing import List, Tuple, Dict, Any, Optional
+import numpy as np
+import logging
+from operator import itemgetter
+import time
+
+from syne_tune.tuner_callback import TunerCallback
+from syne_tune.backend.simulator_backend.simulator_backend import SimulatorBackend
+from syne_tune.backend.trial_status import Trial
+from syne_tune.callbacks.hyperband_remove_checkpoints_score import (
+    compute_probabilities_of_getting_resumed,
+)
+from syne_tune.optimizer.scheduler import SchedulerDecision
+from syne_tune.optimizer.schedulers import HyperbandScheduler
+from syne_tune.optimizer.schedulers.hyperband_stopping import PausedTrialsResult
+
+logger = logging.getLogger(__name__)
+
+
+class TrialStatus:
+    RUNNING = 0
+    PAUSED_WITH_CHECKPOINT = 1
+    PAUSED_NO_CHECKPOINT = 2
+    STOPPED_OR_COMPLETED = 3
+
+
+class BetaBinomialEstimator:
+    """
+    Estimator of the probability :math:`p = P(X = 1)` for a variable :math:`X`
+    with Bernoulli distribution. This is using a Beta prior, which is
+    conjugate to the binomial likelihood. The prior is parameterized by
+    effective sample size ``beta_size`` (:math:`a + b`) and mean ``beta_mean``
+    (:math:`a / (a + b)`).
+    """
+
+    def __init__(self, beta_mean: float, beta_size: float):
+        assert 0 < beta_mean < 1
+        assert beta_size > 0
+        self._beta_a = beta_mean * beta_size
+        self._beta_size = beta_size
+        self._num_one = 0
+        self._num_total = 0
+
+    def update(self, data: List[bool]):
+        self._num_one += sum(data)
+        self._num_total += len(data)
+
+    @property
+    def num_one(self) -> int:
+        return self._num_one
+
+    @property
+    def num_total(self):
+        return self._num_total
+
+    def posterior_mean(self) -> float:
+        a_posterior = self._beta_a + self._num_one
+        aplusb_posterior = self._beta_size + self._num_total
+        return a_posterior / aplusb_posterior
+
+
+class HyperbandRemoveCheckpointsCommon(TunerCallback):
+    """
+    Common base class for :class:`HyperbandRemoveCheckpointsCallback` and
+    :class:`HyperbandRemoveCheckpointsBaselineCallback`.
+    """
+
+    def __init__(
+        self,
+        max_num_checkpoints: int,
```

Comment:
The user only really has to set this one. The other args are filled in by looking at the scheduler and the stopping rule